examples/llm/src/data/finetuning/dataloader.py

```diff
@@ -0,0 +1,287 @@
+        )
+
+        if cfg.dataset.get('packing_ratio'):
+            n_examples_to_pack = int(device_batch_size *
+                                     cfg.dataset.packing_ratio)
+            if n_examples_to_pack < device_batch_size:
+                raise ValueError('packing_ratio must be >= 1, if supplied')
+            if not cfg.dataset.decoder_only_format:
+                raise NotImplementedError(
+                    'On-the-fly packing is currently only supported for decoder-only formats.'
+                )
+            collate_fn = BinPackWrapper(
+                collator=collate_fn,
+                target_batch_size=device_batch_size,
+                max_seq_len=cfg.dataset.max_seq_len,
+                pad_token_id=tokenizer.pad_token_id,
+                padding_side=tokenizer.padding_side,
+                max_leftover_bins_to_keep=cfg.dataset.get(
+                    'max_leftover_bins_to_keep'),
+            )
+            device_batch_size = n_examples_to_pack
+        elif cfg.dataset.get('max_leftover_bins_to_keep') is not None:
+            raise ValueError(
+                'cfg.dataset.max_leftover_bins_to_keep has been defined, ' +\
+                'but cfg.dataset.packing_ratio has not been set. Please set ' +\
+                'the latter to turn on packing or remove the former from the config.')
+
+        return DataLoader(
+            dataset,
+            collate_fn=collate_fn,
+            batch_size=device_batch_size,
+            sampler=dist.get_sampler(dataset,
+                                     drop_last=cfg.drop_last,
+                                     shuffle=cfg.dataset.shuffle),
+            num_workers=cfg.num_workers,
+            pin_memory=cfg.get('pin_memory', True),
+            prefetch_factor=cfg.get('prefetch_factor', 2),
+            persistent_workers=cfg.get('persistent_workers', True),
+            timeout=cfg.get('timeout', 0),
+        )
+
+
+if __name__ == '__main__':
+    import torch
+    from omegaconf import OmegaConf as om
+    cfg = om.create({
+        'dataset': {
+            'name': 'tatsu-lab/alpaca',
+            'split': 'train',
+            'packing_ratio': 18.0,
+            'tokenizer_name': 'gpt2',
+            'max_seq_len': 2048,
+            'decoder_only_format': True,
+            'separator_text': False,
+            'allow_pad_trimming': False,
+            'num_canonical_nodes': 472,
+            'shuffle': True,
+        },
+        'drop_last': False,
+        'num_workers': 0,
+        'pin_memory': False,
+        'prefetch_factor': 2,
+        'persistent_workers': False,
+        'timeout': 0
+    })
+
+    device_batch_size = 8
+
+    dataloader = build_finetuning_dataloader(cfg, device_batch_size)
+
+    tokenizer = transformers.AutoTokenizer.from_pretrained(
+        cfg.dataset.tokenizer_name)
+
+    packing = cfg.dataset.get('packing_ratio') is not None
+
+    for i, batch in enumerate(dataloader):
+        if i >= 100:
```

Comment:
Ah, this was to let the dataloader run through enough batches to get good packing estimates. This is no longer needed because the packing profiling tooling has been added to the packing file itself.
I'll clean this up.